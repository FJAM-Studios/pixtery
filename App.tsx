import {
  NavigationContainer,
  NavigationContainerRef,
} from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import Constants from "expo-constants";
import * as Linking from "expo-linking";
import * as Notifications from "expo-notifications";
import * as SplashScreen from "expo-splash-screen";
import { StatusBar } from "expo-status-bar";
import { requestTrackingPermissionsAsync } from "expo-tracking-transparency";
import * as Updates from "expo-updates";
import React, { useRef, useEffect } from "react";
import { Alert, AppState, LogBox, Dimensions, Platform } from "react-native";
import { GestureHandlerRootView } from "react-native-gesture-handler";
import { Provider as PaperProvider } from "react-native-paper";
import { RootSiblingParent } from "react-native-root-siblings";
import { SafeAreaProvider } from "react-native-safe-area-context";
import {
  Provider as StoreProvider,
  useDispatch,
  useSelector,
} from "react-redux";

import { TabContainer } from "./components/Containers";
import { Splash, TitleScreen } from "./components/TransitionScreens";
import { CreateProfile, EnterName } from "./components/UserScreens";
import { MIN_BOTTOM_CLEARANCE } from "./constants";
import store from "./store";
import { setNotificationToken } from "./store/reducers/notificationToken";
import { setDeviceSize } from "./store/reducers/screenHeight";
import { StackScreens, RootState } from "./types";
import { goToScreen } from "./util";
//less than ideal, but idk if we have a choice right now. suppresses the firebase timeout warning
// LogBox.ignoreLogs(["Setting a timer for a long period of time"]);
LogBox.ignoreLogs([
  "AsyncStorage has been extracted from react-native core and will be removed in a future release",
]);
const Stack = createNativeStackNavigator<StackScreens>();

SplashScreen.preventAutoHideAsync().catch();

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false,
  }),
});

const App = (): JSX.Element => {
  const dispatch = useDispatch();
  const navigationRef = useRef<NavigationContainerRef<StackScreens> | null>(
    null
  );
  const theme = useSelector((state: RootState) => state.theme);

  const promptRestart = () => {
    Alert.alert("A new update is ready. Please restart the app.", "", [
      {
        text: "Close",
        style: "cancel",
      },
      {
        text: "Restart",
        onPress: () => Updates.reloadAsync(),
      },
    ]);
  };
  const getUpdate = async () => {
    try {
      const isUpdate = await Updates.checkForUpdateAsync();
      if (isUpdate.isAvailable) {
        await Updates.fetchUpdateAsync();
        promptRestart();
      }
    } catch (error) {
      console.log(error);
    }
  };

  useEffect(() => {
    //don't check for updates in dev mode
    if (process.env.NODE_ENV !== "development") {
      // when update is downloaded, request reload
      Updates.addListener((event) => {
        if (event.type === Updates.UpdateEventType.UPDATE_AVAILABLE) {
          promptRestart();
        }
      });

      //check for updates when app is foregrounded
      AppState.addEventListener("change", () => {
        if (AppState.currentState === "active") {
          getUpdate();
        }
      });
    }

    async function requestTrackingPermissions() {
      try {
        await requestTrackingPermissionsAsync();
      } catch (error) {
        alert(`trackingErr${error}`);
      }
    }
    requestTrackingPermissions();

    const registerForPushNotificationsAsync = async () => {
      if (Constants.isDevice) {
        const {
          status: existingStatus,
        } = await Notifications.getPermissionsAsync();
        let finalStatus = existingStatus;
        if (existingStatus !== "granted") {
          const { status } = await Notifications.requestPermissionsAsync();
          finalStatus = status;
        }
        if (finalStatus !== "granted") {
          alert("Failed to get push token for push notification!");
          return;
        }
        const token = (
          await Notifications.getExpoPushTokenAsync({
            experienceId: "@fjam-studios/pixtery",
          })
        ).data;
        dispatch(setNotificationToken(token));
      } else {
        alert("Must use physical device for Push Notifications");
      }

      if (Platform.OS === "android") {
        Notifications.setNotificationChannelAsync("default", {
          name: "default",
          importance: Notifications.AndroidImportance.MAX,
          vibrationPattern: [0, 250, 250, 250],
          lightColor: "#FF231F7C",
        });
      }
    };

    registerForPushNotificationsAsync();

    const { width, height } = Dimensions.get("screen");

    const boardSize =
      0.95 *
      Math.min(
        Math.min(height, width),
        MIN_BOTTOM_CLEARANCE * Math.max(height, width)
      );
    dispatch(setDeviceSize(width, height, boardSize));

    // on url change go to the splash screen, which will stop the user if they aren't logged in
    Linking.addEventListener("url", (ev) => {
      const url = ev.url;
      if (url && navigationRef.current)
        goToScreen(navigationRef.current, "Splash", { url });
    });
  }, []);

  // to control trigger order and prevent users from skipping the login screen, puzzle querying has been moved to AddPuzzle, which is called from Splash, which is navigated to only after the navigation container loads using the onReady prop
  const gotoSplash = () => {
    // this timeout is if we want to force users to see the starting screen before moving on.
    if (navigationRef.current) {
      goToScreen(navigationRef.current, "Splash");
    }
  };

  return (
    <PaperProvider theme={theme}>
      <SafeAreaProvider>
        <NavigationContainer ref={navigationRef} onReady={gotoSplash}>
          <Stack.Navigator
            initialRouteName="TitleScreen"
            screenOptions={{ headerShown: false }}
          >
            {/* Initialization Screens */}
            <Stack.Screen name="TitleScreen" component={TitleScreen} />
            <Stack.Screen name="Splash" component={Splash} />

            {/* Login Screens */}
            <Stack.Screen name="CreateProfile" component={CreateProfile} />
            <Stack.Screen name="EnterName" component={EnterName} />

            {/* Tab Container */}
            <Stack.Screen name="TabContainer" component={TabContainer} />
          </Stack.Navigator>
        </NavigationContainer>
      </SafeAreaProvider>
      <StatusBar
        style={theme.dark ? "light" : "dark"}
        backgroundColor={theme.colors.primary}
      />
    </PaperProvider>
  );
};

const AppWrapper = (): JSX.Element => {
  return (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <StoreProvider store={store}>
        <RootSiblingParent>
          <App />
        </RootSiblingParent>
      </StoreProvider>
    </GestureHandlerRootView>
  );
};

export default AppWrapper;
