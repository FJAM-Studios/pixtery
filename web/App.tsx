import { useState, useEffect } from "react";

import { queryPuzzleCallable } from "../FirebaseApp/CloudFunctions";
import { getPixteryURL } from "../FirebaseApp/StorageFunctions";
import { PUBLIC_KEY_LENGTH } from "../constants";
import { Puzzle as PuzzleType } from "../types";
import Puzzle from "./Puzzle";
import StoreLinks from "./StoreLinks";

export default function App(): JSX.Element {
  const [loading, setLoading] = useState(true);
  const [puzzle, setPuzzle] = useState<PuzzleType>();

  useEffect(() => {
    const path = window.location.pathname;
    const publicKey = path.substring(path.lastIndexOf("/") + 1);
    fetchPuzzle(publicKey);
  }, []);

  const fetchPuzzle = async (publicKey: string): Promise<PuzzleType | void> => {
    if (publicKey.length === PUBLIC_KEY_LENGTH) {
      try {
        const puzzleData = (await queryPuzzleCallable({ publicKey }))
          .data as Puzzle;
        const imageURI = puzzleData.imageURI;
        const downloadURL = await getPixteryURL("/" + imageURI);
        puzzleData.imageURI = downloadURL;
        setPuzzle(puzzleData);
      } catch (error) {
        console.error(error);
      }
    }
    setLoading(false);
  };

  if (loading) {
    return (
      <div id="app">
        <h1 style={{ fontFamily: "Shrikhand" }}>Loading...</h1>
      </div>
    );
  }
  if (puzzle) {
    return (
      <div id="app">
        <Puzzle puzzle={puzzle} />
      </div>
    );
  }
  return (
    <div id="app">
      <img src="/logo.svg" className="logo" alt="Pixtery!" />
      <img src="/pixtery.svg" className="title" alt="Pixtery!" />
      <StoreLinks />
      <div id="web-footer">
        <div id="web-footer-contents">
          <a href="https://pixtery.io/privacy.html">Privacy Policy</a>
          <span className="dot">&nbsp;&#x2022;&nbsp;</span>
          <a href="https://pixtery.io/community.html">Community Guidelines</a>
          <span className="dot">&nbsp;&#x2022;&nbsp;</span>
          <a href="https://pixtery.io/terms.html">Terms</a>
          <span className="dot">&nbsp;&#x2022;&nbsp;</span>
          <a href="mailto:contact@pixtery.io">contact@pixtery.io</a>
          <div>
            <span className="studio">Â© 2021 FJAM Studios</span>
          </div>
        </div>
      </div>
    </div>
  );
}
