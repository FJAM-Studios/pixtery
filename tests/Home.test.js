import React from 'react';
import renderer from 'react-test-renderer';
import App from '../App.tsx';

test('Home screen renders correctly', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});
