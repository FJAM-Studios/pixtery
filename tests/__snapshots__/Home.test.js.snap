// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen renders correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  />
</View>
`;
